File: Java/src/_0968_Binary_Tree_Cameras/Solution.java
package _0968_Binary_Tree_Cameras;

/**
 * Date: 2021/7/8 22:04
 * Content:
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：38.2 MB, 在所有 Java 提交中击败了40.37%的用户
 */
public class Solution {

    static int INF = (int) (1e8);

    public int minCameraCover(TreeNode root) {
        int[] f = dfs(root);
        return Math.min(f[1], f[2]);
    }

    int[] dfs(TreeNode root) {
        if (root == null) return new int[]{0, 0, INF};
        int[] l = dfs(root.left), r = dfs(root.right);
        return new int[]{
                Math.min(l[1], l[2]) + Math.min(r[1], r[2]),
                Math.min(l[2] + Math.min(r[1], r[2]), r[2] + Math.min(l[1], l[2])),
                Math.min(l[0], Math.min(l[1], l[2])) + Math.min(r[0], Math.min(r[1], r[2])) + 1,
        };
    }

    public static void main(String[] args) {

        int nu = Integer.MIN_VALUE;  // 用Integer.MIN_VALUE表示二叉树 null
        int[] nums = new int[]{
                1,
                1, nu,
                1, 1,
        };
        TreeNode root = (new MyTree(nums)).getRoot();
        System.out.println((new Solution()).minCameraCover(root));  // 1
    }
}

File: Python/_0968_Binary_Tree_Cameras/Solution.py
# coding=utf-8
# Date: 2021/7/8 22:16
from MyTree import *


# 执行用时：48 ms, 在所有 Python3 提交中击败了91.95%的用户
# 内存消耗：15.3 MB, 在所有 Python3 提交中击败了30.65%的用户
class Solution:
    def minCameraCover(self, root: TreeNode) -> int:
        f = self.dfs(root)
        return min(f[1], f[2])

    def dfs(self, root):
        if root is None:
            return 0, 0, int(1e8)
        l, r = self.dfs(root.left), self.dfs(root.right)
        return (
            min(l[1], l[2]) + min(r[1], r[2]),
            min(l[2] + min(r[1], r[2]), r[2] + min(l[1], l[2])),
            min(l[0], min(l[1], l[2])) + min(r[0], min(r[1], r[2])) + 1,
        )


if __name__ == "__main__":
    nu = None
    nums = [
        1,
        1, nu,
        1, 1,
    ]
    root = MyTree(nums).getRoot()
    print(Solution().minCameraCover(root))  # 1
